ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java

```diff
@@ -77,86 +88,138 @@ public AmbryQuotaManager(QuotaConfig quotaConfig, QuotaRecommendationMergePolicy
   }
 
   @Override
-  public void init() throws InstantiationException {
-    Timer.Context timer = quotaMetrics.quotaManagerInitTime.time();
-    try {
-      for (QuotaEnforcer quotaEnforcer : requestQuotaEnforcers) {
-        quotaEnforcer.init();
-      }
-    } catch (Exception e) {
-      logger.error("Failed to init quotaEnforcer", e);
-      if (e instanceof InstantiationException) {
-        throw (InstantiationException) e;
-      } else {
-        throw new InstantiationException(e.getMessage());
+  public synchronized void init() throws Exception {
+    if (!initialized.get()) {
+      Timer.Context timer = quotaMetrics.quotaManagerInitTime.time();
+      try {
+        for (QuotaEnforcer quotaEnforcer : quotaEnforcers) {
+          quotaEnforcer.init();
+        }
+      } catch (Exception e) {
+        LOGGER.error("Failed to init quotaEnforcer", e);
+        throw e;
+      } finally {
+        timer.stop();
       }
-    } finally {
-      timer.stop();
+      LOGGER.info("AmbryQuotaManager initialized.");
+      initialized.set(true);
+    } else {
+      LOGGER.warn("AmbryQuotaManager already initialized.");
     }
   }
 
+  /**
+   * {@inheritDoc}
+   *
+   * If there are any {@link QuotaEnforcer#recommend} calls that throw exception, this implementation logs the exception,
+   * and makes a best effort to gather as many recommendations as possible. Recommendations for some {@link QuotaName}s
+   * will be missing in those cases.
+   */
   @Override
-  public ThrottlingRecommendation getThrottleRecommendation(RestRequest restRequest) {
-    if (requestQuotaEnforcers.isEmpty()) {
+  public ThrottlingRecommendation recommend(RestRequest restRequest) throws QuotaException {
+    quotaMetrics.recommendRate.mark();
+    if (quotaEnforcers.isEmpty()) {
+      LOGGER.warn("No quota enforcers in AmbryQuotaManager. Returning null recommendation.");
+      quotaMetrics.noQuotaRecommendationRate.mark();
       return null;
     }
     ThrottlingRecommendation throttlingRecommendation;
-    Timer.Context timer = quotaMetrics.quotaEnforcementTime.time();
+    Timer.Context timer = quotaMetrics.quotaRecommendationTime.time();
     try {
       List<QuotaRecommendation> quotaRecommendations = new ArrayList<>();
-      for (QuotaEnforcer quotaEnforcer : requestQuotaEnforcers) {
+      boolean partialRecommendation = false;
+      boolean noRecommendation = true;
```

Comment:
Nit: These can be inferred from the size of quotaRecommendations

quotaRecommendations.empty() => no recommendation
quotaRecommendations.size() < quotaEnforcers.size() => partial recommendation

Reply:
fixed.